sql/index/pilosa/driver_test.go

```diff
@@ -0,0 +1,234 @@
+
+	it := &testIndexKeyValueIter{
+		offset:      0,
+		total:       128,
+		expressions: expressions,
+	}
+
+	err = d.Save(context.Background(), sqlIdx, it)
+	require.Nil(err)
+
+	indexes, err := d.LoadAll(db, table)
+	require.Nil(err)
+	require.Equal(1, len(indexes))
+	require.Truef(reflect.DeepEqual(sqlIdx, indexes[0]), "Not equal indexes")
+
+	for _, r := range it.records {
+		lookup, err := sqlIdx.Get(r.values...)
+		require.Nil(err)
+
+		found, foundLoc := false, []string{}
+		lit := lookup.Values()
+		for i := 0; ; i++ {
+			loc, err := lit.Next()
+			t.Logf("[%d] values: %v location: %x loc: %x err: %v\n", i, r.values, r.location, loc, err)
+
+			if err == io.EOF {
+				require.Truef(i > 0, "No data for r.values: %v\tr.location: %x", r.values, r.location)
+				break
+			}
+
+			require.Nil(err)
+			found = found || reflect.DeepEqual(r.location, loc)
+			foundLoc = append(foundLoc, hex.EncodeToString(loc))
+		}
+		require.Truef(found, "Expected: %s\nGot: %v\n", hex.EncodeToString(r.location), foundLoc)
+
+		err = lit.Close()
+		require.Nil(err)
+	}
+}
+
+func TestSaveAndGetAll(t *testing.T) {
+	require := require.New(t)
+
+	db, table, id := "db_name", "table_name", "index_id"
+	path, err := mkdir(os.TempDir(), "indexes")
+	require.Nil(err)
+	defer os.RemoveAll(path)
+
+	h1 := sha1.Sum([]byte("lang"))
+	exh1 := sql.ExpressionHash(h1[:])
+
+	h2 := sha1.Sum([]byte("hash"))
+	exh2 := sql.ExpressionHash(h2[:])
+
+	expressions := []sql.ExpressionHash{exh1, exh2}
+
+	d := NewDriver(path, pilosa.DefaultClient())
+	sqlIdx, err := d.Create(db, table, id, expressions, nil)
+	require.Nil(err)
+
+	it := &testIndexKeyValueIter{
+		offset:      0,
+		total:       128,
+		expressions: expressions,
+	}
+
+	err = d.Save(context.Background(), sqlIdx, it)
+	require.Nil(err)
+
+	indexes, err := d.LoadAll(db, table)
+	require.Nil(err)
+	require.Equal(1, len(indexes))
+	require.Truef(reflect.DeepEqual(sqlIdx, indexes[0]), "Not equal indexes")
+
+	lookup, err := sqlIdx.Get()
+	require.Nil(err)
+
+	lit := lookup.Values()
+	for i := 0; ; i++ {
+		loc, err := lit.Next()
+		if err == io.EOF {
+			break
+		}
+		require.Nil(err)
+
+		t.Logf("[%d] values: %v location: %x loc: %x err: %v\n", i, it.records[i].values, it.records[i].location, loc, err)
+		require.Truef(reflect.DeepEqual(it.records[i].location, loc),
+			"Expected: %v\nGot: %v\n", it.records[i].location, loc)
+	}
+
+	err = lit.Close()
+	require.Nil(err)
+}
+
+func TestDelete(t *testing.T) {
+	require := require.New(t)
+
+	db, table, id := "db_name", "table_name", "index_id"
+	path, err := mkdir(os.TempDir(), "indexes")
+	require.Nil(err)
+	defer os.RemoveAll(path)
+
+	h1 := sha1.Sum([]byte("lang"))
+	exh1 := sql.ExpressionHash(h1[:])
+
+	h2 := sha1.Sum([]byte("hash"))
+	exh2 := sql.ExpressionHash(h2[:])
+
+	expressions := []sql.ExpressionHash{exh1, exh2}
+
+	d := NewIndexDriver(path)
+	sqlIdx, err := d.Create(db, table, id, expressions, nil)
+	require.Nil(err)
+
+	err = d.Delete(sqlIdx)
+	require.Nil(err)
+}
```

Comment:
Can we test what happens if the index generation is stopped abruptly?

Reply:
We can add some tests for "failing cases", but generally with docker container after restart it will recreate a schema. At the end we won't find anything in bitmaps. The only functions which should work are `Has` and `Get()` (empty get) where we do not touch pilosa.
So what's your opinion?

Reply:
Have some tests for failing cases sound great. I'm afraid of having an index loaded that make a query fails just because is corrupted, or an index that appears to work correctly but it has incorrect or partial results.